common/ast/src/parser/query.rs

```diff
@@ -305,3 +271,131 @@ pub fn order_by_expr(i: Input) -> IResult<OrderByExpr> {
         },
     )(i)
 }
+
+pub fn select(i: Input) -> IResult<SetExpr> {
+    map(
+        consumed(rule! {
+             SELECT ~ DISTINCT? ~ #comma_separated_list1(select_target)
+                ~ ( FROM ~ ^#comma_separated_list1(table_reference) )?
+                ~ ( WHERE ~ ^#expr )?
+                ~ ( GROUP ~ ^BY ~ ^#comma_separated_list1(expr) )?
+                ~ ( HAVING ~ ^#expr )?
+        }),
+        |(
+            span,
+            (
+                _select,
+                opt_distinct,
+                select_list,
+                opt_from_block,
+                opt_where_block,
+                opt_group_by_block,
+                opt_having_block,
+            ),
+        )| {
+            SetExpr::Select(Box::new(SelectStmt {
+                span: span.0,
+                distinct: opt_distinct.is_some(),
+                select_list,
+                from: opt_from_block
+                    .map(|(_, table_refs)| table_refs)
+                    .unwrap_or_default(),
+                selection: opt_where_block.map(|(_, selection)| selection),
+                group_by: opt_group_by_block
+                    .map(|(_, _, group_by)| group_by)
+                    .unwrap_or_default(),
+                having: opt_having_block.map(|(_, having)| having),
+            }))
+        },
+    )(i)
+}
+
+pub fn parenthesized_select(i: Input) -> IResult<SetExpr> {
+    map(
+        rule! (
+            "(" ~ (#parenthesized_select  | #select ) ~ ")"
```

Comment:
You can write this as:

```
( "(" ~ #parenthesized_select ~ ")" ) | #select
```

Both parenthesized `SELECT` statement and `SELECT` statement can be reduced into `SetExpr` with rule.

So you don't have to write `(#select | #parenthesized_select)` in https://github.com/datafuselabs/databend/blob/1bdac9f4eaee2eeee13df184bec6206569ad918d/common/ast/src/parser/query.rs#L325

Reply:
> ( "(" ~ #parenthesized_select ~ ")" ) | #select

`( "(" ~ #parenthesized_select ~ ")" )` and `#select` don't have same type, so can we combine them with `|`?